core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -253,7 +253,7 @@ private[spark] class Executor(
 
         val directResult = new DirectTaskResult(valueBytes, accumUpdates, task.metrics.orNull)
         val serializedDirectResult = ser.serialize(directResult)
-        val resultSize = serializedDirectResult.limit
+        val resultSize = serializedDirectResult.remaining()
```

Comment:
> Do we have a situation where the position is not 0 though, but is correctly at the start of the data?

If a `ByteBuffer` is from Netty, the position could be a non-zero value.

> Equally, if that's an issue, are we sure the entire buffer has valid data, through the end? that assumption is still present here, that the end of the data is the end of the buffer.

The `ByteBuffer` may contain more data internally, but the user should only read the part between `position` and `limit`. I think that's defined in `ByteBuffer/Buffer` javadoc.